python/cuml/manifold/t_sne.pyx

```diff
@@ -339,6 +340,9 @@ class TSNE(Base,
             raise ValueError("post_momentum = {} should be more than "
                              "pre_momentum = {}".format(post_momentum,
                                                         pre_momentum))
+        if method == "barnes_hut":
+            warnings.warn("Starting from version 22.02, the default method "
```

Comment:
From looking at https://github.com/rapidsai/cuml/issues/3805, the gap between BH and FFT appears to be pretty small. Before we "officially" promote the FFT method to the default, however, we should consider doing a couple scale tests (preferably over 1M) but otherwise the perf price we pay here really doesn't look bad at all for the gain in stability. 

We should address this comment, though, at the very least: https://github.com/rapidsai/cuml/issues/3865#issuecomment-841940453

Reply:
One thing I just noticed in the code is that the early stopping seems to have gotten removed from the barnes-hut implementaiton at some point so it looks like it'll always iterate to max_iter. This is going to make it hard to get an end-to-end benchmark that compares w/ fft. It shouldn't be hard to add this back in.

Reply:
Yes I didn't modify this parameter when doing the benchmarks

Reply:
And for the comment that you linked I was able to run the datasets provided without seeing significant decrease in performance

Reply:
@lowener, the runtimes of the two algorithms aren't comparable when one algorithm supports early stopping and the other doesn't. [Here](https://github.com/rapidsai/cuml/blob/bb588e9f208be0f7a3e165b14954633be89aac6a/cpp/src/tsne/fft_tsne.cuh#L579) is where we perform the early stopping in the FFT implementation. This used to be in the BH implementation as well but it looks like it got removed at some point. While the FFT implementation is slower iteration for iteration, we expect it should be converging faster than barnes-hut (should be stopping earlier). We can't compare the two in benchmarks when barnes-hut isn't able to stop early.